File: src/main/java/pl/ogiba/dwarf/scenes/MainPresenter.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package pl.ogiba.dwarf.scenes;

import com.mongodb.MongoClient;
import com.mongodb.MongoClientOptions;
import com.mongodb.ServerAddress;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.MongoIterable;
import com.mongodb.event.ServerClosedEvent;
import com.mongodb.event.ServerHeartbeatFailedEvent;
import com.mongodb.util.JSON;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import org.bson.Document;
import pl.ogiba.dwarf.utils.ServerListenerAdapter;
import pl.ogiba.dwarf.utils.ServerMonitorListenerAdapter;

/**
 *
 * @author ogiba
 */
public class MainPresenter implements IMainPresenter {

    private final IMainView mainView;

    private MongoClient client;
    private MongoDatabase db;
    private final ArrayList<String> loadedDBs;

    private boolean isConnected = false;

    public MainPresenter(IMainView mainView) {
        this.mainView = mainView;
        this.loadedDBs = new ArrayList<>();
    }

    @Override
    public void checkConnection() {
        if (isConnected) {
            disconnect();
        } else {
            connectToDb();
        }
    }

    @Override
    public void loadData() {
        if (db == null || !isConnected) {
            return;
        }

        final String dbName = db.getName();

        if (!loadedDBs.contains(dbName)) {
            this.loadedDBs.add(dbName);
            mainView.onDataLoaded(dbName);
        }
    }

    @Override
    public void loadCollections() {
        if (db == null || !isConnected) {
            return;
        }

        MongoIterable<String> collections = db.listCollectionNames();
        ArrayList<String> values = collections.into(new ArrayList<>());
        System.err.println(String.format("Number of collections: %s", values.size()));

        mainView.onCollectionsLoaded(values);
    }

    @Override
    public void loadSelectedCollection(String collection) {
        MongoCollection<Document> selectedCollection = db.getCollection(collection);

        System.out.println(selectedCollection.find().first());

        String data = JSON.serialize(selectedCollection.find());
        mainView.onSelectedCollectionLoaded(data);
    }

    private void connectToDb() {
        MongoClientOptions clientOptions = new MongoClientOptions.Builder()
                .addServerListener(new ServerListenerAdapter() {
                    @Override
                    public void serverClosed(ServerClosedEvent event) {
                        isConnected = false;
                        mainView.onConnectionResult(isConnected);
                    }
                })
                .addServerMonitorListener(new ServerMonitorListenerAdapter() {
                    @Override
                    public void serverHeartbeatFailed(ServerHeartbeatFailedEvent event) {
                        super.serverHeartbeatFailed(event);

                        isConnected = false;
                        mainView.onConnectionResult(isConnected);
                    }

                })
                .build();

        ServerAddress serverAddress = new ServerAddress("127.0.0.1", 27017);

        client = new MongoClient(serverAddress, clientOptions);

        db = client.getDatabase("dwarf_test");

        isConnected = true;

        mainView.onConnectionResult(isConnected);
    }

    private void disconnect() {
        client.close();
        client = null;
        db = null;
    }
}

File: src/main/java/pl/ogiba/dwarf/scenes/main/MainPresenter.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package pl.ogiba.dwarf.scenes.main

import com.mongodb.MongoClient
import com.mongodb.MongoClientOptions
import com.mongodb.ServerAddress
import com.mongodb.client.MongoCollection
import com.mongodb.client.MongoDatabase
import com.mongodb.event.ServerClosedEvent
import com.mongodb.event.ServerHeartbeatFailedEvent
import com.mongodb.util.JSON
import org.bson.Document
import pl.ogiba.dwarf.utils.ServerListenerAdapter
import pl.ogiba.dwarf.utils.ServerMonitorListenerAdapter
import java.util.*

/**
 *
 * @author ogiba
 */
class MainPresenter(private val mainView: IMainView) : IMainPresenter {

    private var client: MongoClient? = null
    private var db: MongoDatabase? = null
    private val loadedDBs: MutableList<String> = mutableListOf()
    private var selectedCollection: MongoCollection<Document>? = null

    private var isConnected = false

    override fun checkConnection() {
        if (isConnected) {
            disconnect()
        } else {
            connectToDb()
        }
    }

    override fun loadData() {
        db?.takeIf { isConnected }?.run {
            loadedDBs.takeIf { !it.contains(name) }?.run {
                add(name)
                mainView.onDataLoaded(name)
            }
        }
    }

    override fun loadCollections() {
        db?.takeIf { isConnected }?.run {
            val collections = listCollectionNames()
            val values = collections.into(ArrayList())
            System.err.println(String.format("Number of collections: %s", values.size))
            mainView.onCollectionsLoaded(values)
        }
    }

    override fun loadSelectedCollection(collection: String) {
        mainView.onColectionSelected()
        selectedCollection = db?.getCollection(collection)?.also { selected ->
            println(selected.find().first())
            mainView.onSelectedCollectionLoaded(JSON.serialize(selected.find()))
        }
    }

    override fun proceedInsertAction() {
        mainView.onInsertApplied(selectedCollection!!)
    }

    private fun connectToDb() {
        val clientOptions = MongoClientOptions.Builder().addServerListener(object : ServerListenerAdapter() {
            override fun serverClosed(event: ServerClosedEvent) {
                isConnected = false
                mainView.onConnectionResult(isConnected)
            }
        }).addServerMonitorListener(object : ServerMonitorListenerAdapter() {
            override fun serverHeartbeatFailed(event: ServerHeartbeatFailedEvent) {
                super.serverHeartbeatFailed(event)

                isConnected = false
                mainView.onConnectionResult(isConnected)
            }

        }).build()

        val serverAddress = ServerAddress("127.0.0.1", 27017)

        client = MongoClient(serverAddress, clientOptions).also { newClient ->
            db = newClient.getDatabase("dwarf_test")
            isConnected = true
        }
        mainView.onConnectionResult(isConnected)
    }

    private fun disconnect() {
        client?.close()
        client = null
        db = null
    }
}
